kore/src/Kore/Step/Implication.hs

```diff
@@ -0,0 +1,452 @@
+        )
+  where
+    Implication { right, existentials } = implication'
+
+freeVariablesLeft
+    :: Implication modality
+    -> FreeVariables RewritingVariableName
+freeVariablesLeft implication'@(Implication _ _ _ _ _) =
+    freeVariables left
+  where
+    Implication { left } = implication'
+
+instance HasFreeVariables (Implication modality) RewritingVariableName where
+    freeVariables implication'@(Implication _ _ _ _ _) =
+        freeVariablesLeft implication'
+        <> freeVariablesRight implication'
+
+-- | Creates an 'Implication' from a left hand side 'Pattern'
+-- and an 'OrPattern', representing the right hand side pattern.
+-- The list of element variables are existentially quantified
+-- in the right hand side.
+mkImplication
+    :: modality
+    -> Pattern RewritingVariableName
+    -> OrPattern RewritingVariableName
+    -> [ElementVariable RewritingVariableName]
+    -> Implication modality
+mkImplication modality left right existentials =
+    Implication
+        { left
+        , right
+        , modality
+        , existentials
+        , attributes = Default.def
+        }
+
+{- | Construct a 'TermLike' from the parts of an implication-based rule.
+
+The 'TermLike' has the following form:
+
+@
+\\implies{S}(\\and{S}(left, requires), alias{S}(right))
+@
+
+that is,
+
+@
+left ∧ requires → alias(right)
+@
+
+ -}
+implicationToTerm
+    :: Modality
+    -> Implication modality
+    -> TermLike VariableName
+implicationToTerm modality representation@(Implication _ _ _ _ _) =
+    TermLike.mkImplies
+        (TermLike.mkAnd leftCondition leftTerm)
+        (TermLike.applyModality modality rightPattern)
+  where
+    Implication { left, right, existentials } = representation
+    leftTerm =
+        Pattern.term left
+        & getRewritingTerm
+    sort = TermLike.termLikeSort leftTerm
+    leftCondition =
+        Pattern.withoutTerm left
+        & Pattern.fromCondition sort
+        & Pattern.toTermLike
+        & getRewritingTerm
+    rightPattern =
+        TermLike.mkExistsN existentials (OrPattern.toTermLike right)
+        & getRewritingTerm
+
+substituteRight
+    :: Map
+        (SomeVariableName RewritingVariableName)
+        (TermLike RewritingVariableName)
+    -> Implication modality
+    -> Implication modality
+substituteRight rename implication'@Implication{ right, existentials } =
+    implication'
+        { right = OrPattern.substitute newSubst right
+        , existentials = renameVariable <$> existentials
+        }
+  where
+    existentials' = inject . TermLike.variableName <$> existentials
+    subst = foldr Map.delete rename existentials'
+    targetVars = Foldable.foldl' Set.union Set.empty
+        (FreeVariables.toNames . freeVariables @_ @RewritingVariableName
+        <$> subst)
+    freeVars = FreeVariables.toNames
+        (foldMap freeVariables (OrPattern.toPatterns right))
+        Set.\\ Map.keysSet subst
+        Set.\\ Set.fromList existentials'
```

Comment:
These are the free variables of the RHS of the implication (`right`) before the substitution.
We use these along with `targetVars` to build the set of variables that we need to avoid when renaming the existentials, so that we don't capture variables that were free to begin with or stuff in the terms that get substituted in.
We're subtracting the set of keys of `subst` from this because there's no problem if the existentials get renamed to those (their initial free occurences will get replaced)